### DISCUSSION TITLE
parallel_for equivalent in Ginkgo

### DISCUSSION THREAD

**Discussion by Timo-Schrader** [ginkgo-project/ginkgo#1190] [Q&A] [answered]
Hello,

we have the following (mathematical) problem:

We want to setup a RBF system matrix in order to solve for the coefficients. We furthermore want to accelerate this using CUDA (Gingko) by not only solving the linear system of equations, but also setting up the kernel matrix on the GPU.

Now our question is if there is a way in Ginkgo (similar to parallel_for in Kokkos) to fill such a matrix on the GPU, i.e., evaluate each RBF entry on the GPU side **without** having to write custom CUDA kernels ourselves and pass an array pointer to it. It might be the case that we just didn't find this feature yet, so please excuse me if this question is already answered.

Thank you and have a great day!

Best regards

**Comment by upsj** [comment, marked as answer]
We have such a framework in place already, it's only not yet exposed to the user since we are not certain it is final yet. For an example, see https://github.com/ginkgo-project/ginkgo/pull/938, or more specifically [stencil_kernel.cpp](https://github.com/ginkgo-project/ginkgo/blob/a195f856ee946b1b9387d30513dd992df4501df3/examples/custom-matrix-format/stencil_kernel.cpp). I'd be happy to pick this PR up again if it's of importance to you.

**Reply by Timo-Schrader** [reply]
Thank you for the quick reply! I am not sure yet if this fits our needs. We basically want to use RBF kernels (e.g., Gaussian) and evaluate them on the GPU for each input vertex (before solving the entire system). However, it seems to me that the kernel that you linked, assumes that these values had been calculated already. Or is it possible to also provide a lambda function for example which is then evaluated for every data point independently too?

**Reply by tcojean** [reply]
In the example above, the `GKO_KERNEL` part is a lambda function that does the computation, so you could write your own like so. Note that this system isn't nearly as advanced as what Kokkos or RAJA provide, so depending on your actual needs you might have a much better time using these specialized frameworks instead.

See: https://github.com/ginkgo-project/ginkgo/blob/a195f856ee946b1b9387d30513dd992df4501df3/examples/custom-matrix-format/stencil_kernel.cpp#L52-L62

**Reply by Timo-Schrader** [reply]
That makes sense, thank you very much. I will look at this example and test if it satisfies our needs.

**Reply by davidscn** [reply]
Sorry for butting in: May I ask you to elaborate a bit more in detail what you mean by

> Note that this system isn't nearly as advanced as what Kokkos or RAJA provide

Are you referring to the (not yet final) user interface and corresponding use-cases or the back end implementation?

**Reply by tcojean** [reply]
I was mostly talking about the back end here which is much less advanced than for Kokkos/RAJA. We do not integrate e.g. atomics and several other important features. This generic kernel implementation was made for simple kernels first and foremost, for which we can get good performance with a simple parallelization scheme and we have little to gain by manually writing separate kernels for every hardware backend (CUDA, HIP, SYCL, OpenMP, ...). If the kernels you have in mind fit this scheme, then they can indeed be implemented with this interface.

**Reply by upsj** [reply]
Basically, what we provide right now is nothing more than a `parallel_for`-like dispatch mechanism for all backends, with some mapping functionality for turning Ginkgo types into equivalent device representations. The kernels can operate pointwise on an 1D range or a 2D grid, and we provide reduction operations for things like sums, min/max etc. either along dimensions or over the whole grid. This is very purpose-built for implementing some internal functionality (BLAS-1/2 type operations, format conversions, ...), but may not be sufficient for applications where you might need nested parallelism, inter-thread communication etc.

**Reply by davidscn** [reply]
Got it, thanks a lot for your quick replies!

**Reply by Timo-Schrader** [reply]
Hello again,

so I tried to implement a basic workflow in both Kokkos and Ginkgo and it seems to me that Ginkgo is **much** easier to use as a math library than Kokkos. I was also able to launch a self-written CUDA kernel with the help of Ginkgo using the experimental feature branch. 

In summary, I really like this feature and would be delighted if it is possible to pick up this PR again (as suggested).
Thank you very much for your efforts!

**Reply by upsj** [reply]
This is great to hear - we will discuss how to move this forward again. If you have a minute, could I get some input on https://github.com/ginkgo-project/ginkgo/issues/1209 from the user side? The duplication of arguments/parameters is currently necessary because we need to transform host objects to device representations, but I was wondering how much of an issue potential parameter order/name mismatches actually are.

**Reply by Timo-Schrader** [reply]
After getting used to it/understanding how it works, I think the way it currently works is absolutely feasible for the user as long as it is well documented since "all the magic" happens in `run_kernel.`
However, looking at #1209 , I am not 100% sure what `row_vector(x)` and `default_stride` refer to. 

Thank you very much for your efforts, much appreciated!

**Reply by Timo-Schrader** [reply]
Good afternoon,

we just wanted to report that we are getting first very promising results using Ginkgo + CUDA in [preCICE](https://github.com/precice/precice) when we need to solve a very large RBF system for data mapping. So if you need some further input from our side to get this really great feature of kernel dispatches into the main branch of Ginkgo just let us know!
